### DISCUSSION TITLE
App and MainWindow Destructors Not Being Called

### DISCUSSION THREAD

**Discussion by hamu777** [microsoft/microsoft-ui-xaml#10102] [Q&A] [answered]
When the WinUI3 application exits, the destructors for both the App class and MainWindow are not being called. Is this the expected behavior?
I confirmed that the MainWindow destructor gets executed when I set the window member to nullptr within the App class (C++). How can I ensure that the App class destructor is properly called? This issue occurs in both C++ and C#.

You can easily reproduce this issue by adding a destructor to the App class in a new project.

```C++
    struct App : AppT<App>
    {
        App();

        ~App()
        {
            // Destructor
        }
```

**Comment by DarranRowe** [comment, marked as answer]
As WinUI 3 is currently implemented, the Xaml runtime will never release the last reference it holds to the application object. So yes, it is expected that the destructors for App and any objects held in App will not be called. There is no requirement in Windows for an application to explicitly release all memory used back to Windows as the operating system reclaims all of this memory when the process is exited.
There is no correct way to force App to be destroyed when the application exits, but it is possible to customise the main function and get App to release any objects that it owns.

**Reply by hamu777** [reply]
Thank you for your explanation. I understand that there is no need to explicitly force the App object to be destroyed when the application exits. Thank you for the helpful information you shared.
